### PR TITLE
Refactor get topic names and types

### PR DESCRIPTION
**Pull request by wjwwood**
Connects to ros2/ros2#361

### CODE REVIEW

File: rmw_fastrtps_cpp/src/functions.cpp [hunk 1]
```diff
@@ -2882,12 +3122,19 @@ rmw_get_node_names(
...
+    return rmw_convert_rcutils_ret_to_rmw_ret(rcutils_ret);
+  }
   for (size_t i = 0; i < participant_names.size(); ++i) {
-    size_t name_length = participant_names[i].size() + 1;
-    node_names->data[i] = static_cast<char *>(rmw_allocate(name_length * sizeof(char)));
-    snprintf(node_names->data[i], name_length, "%s", participant_names[i].c_str());
+    node_names->data[i] = rcutils_strdup(participant_names[i].c_str(), allocator);
+    if (!node_names->data[i]) {
+      RMW_SET_ERROR_MSG("failed to allocate memory for node name")
+      return RMW_RET_BAD_ALLOC;
```
Comment (dirk-thomas) [inline rmw_fastrtps_cpp/src/functions.cpp]: How are previously allocated `data[i]` freed in this case? Does it need something like `fail_cleanup` above?

File: rmw_fastrtps_cpp/src/functions.cpp [hunk 2]
```diff
@@ -461,13 +466,8 @@ class ReaderInfo : public ReaderListener
...
     if (change->kind == ALIVE) {
-      if (
-        it == topicNtypes.end() ||
-        it->second.find(proxyData.typeName()) == it->second.end())
-      {
-        topicNtypes[fqdn].insert(proxyData.typeName());
-        trigger = true;
-      }
+      topicNtypes[fqdn].insert(proxyData.typeName());
+      trigger = true;
```
Comment (dirk-thomas) [inline rmw_fastrtps_cpp/src/functions.cpp]: With `it` being removed from the `if` block the declaration of `auto it = ...` can be moved into the `else` block.

Same below.

Comment (wjwwood) [inline rmw_fastrtps_cpp/src/functions.cpp]: Sure: ba254ee

File: rmw_fastrtps_cpp/src/functions.cpp (hunk 1 above)
Comment (wjwwood) [inline rmw_fastrtps_cpp/src/functions.cpp]: See: bfa8aab I didn't use a label because this was at the end of the function anyways.

**Pull request merged**